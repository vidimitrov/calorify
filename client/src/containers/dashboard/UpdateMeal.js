import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from '@reach/router';
import { isEmpty } from 'lodash';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import ArrowBack from '@material-ui/icons/ArrowBack';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import CustomSnackbar from '../../components/common/CustomSnackbar';
import Form from '../../components/dashboard/Form';
import FormControls from '../../components/dashboard/FormControls';
import InputWrapper from '../../components/dashboard/InputWrapper';
import Logo from '../../components/dashboard/Logo';
import Wrapper from '../../components/dashboard/Wrapper';
import CustomTypography from '../../components/dashboard/CustomTypography';
import CustomIconButton from '../../components/dashboard/CustomIconButton';
import CustomTextField from '../../components/dashboard/CustomTextField';
import logo from '../../assets/img/logo.png';
import {
  fetchMeals as fetchMealsActionCreator,
  updateMeal as updateMealActionCreator,
} from '../../actions/meals';

export class UpdateMeal extends React.Component {
  constructor(props) {
    super(props);

    const { meals, mealId } = props;
    const meal = meals.find(m => m.id === mealId);

    this.state = {
      name: meal ? meal.name : '',
      calories: meal ? meal.calories : '',
      date: meal ? meal.date : '',
      time: meal ? meal.time : '',
      change: false,
      positiveSnackbarOpen: false,
      negativeSnackbarOpen: false,
      positiveMessage: null,
      negativeMessage: null,
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.handlePositiveSnackbarClose = this.handlePositiveSnackbarClose.bind(this);
    this.handleNegativeSnackbarClose = this.handleNegativeSnackbarClose.bind(this);
  }

  async componentDidMount() {
    const { meals, getAllMeals, mealId } = this.props;

    if (meals.length === 0) {
      try {
        const fetchedMeals = await getAllMeals();
        const meal = fetchedMeals.find(m => m.id === mealId);

        this.setState({
          name: meal ? meal.name : '',
          calories: meal ? meal.calories : '',
          date: meal ? meal.date : '',
          time: meal ? meal.time : '',
        });
      } catch (error) {
        this.setState({
          negativeSnackbarOpen: true,
          negativeMessage: 'Something went wrong!',
        });
      }
    }
  }

  onChangeHandler(key, value) {
    if (key === 'datetime') {
      this.setState({
        date: value.split('T')[0],
        time: value.split('T')[1],
      });
    } else {
      this.setState({
        change: true,
        [key]: value,
      });
    }
  }

  handlePositiveSnackbarClose(event, reason) {
    if (reason === 'clickaway') {
      return;
    }

    this.setState({
      positiveSnackbarOpen: false,
      positiveMessage: null,
    });
  }

  handleNegativeSnackbarClose(event, reason) {
    if (reason === 'clickaway') {
      return;
    }

    this.setState({
      negativeSnackbarOpen: false,
      negativeMessage: null,
    });
  }

  render() {
    const {
      user,
      mealId,
      updateMeal,
    } = this.props;
    const {
      name,
      calories,
      date,
      time,
      change,
      positiveSnackbarOpen,
      negativeSnackbarOpen,
      positiveMessage,
      negativeMessage,
    } = this.state;
    const { navigate } = this.props;

    if (!user) {
      return (
        <Redirect to="/auth/login" />
      );
    }

    return (
      <Wrapper data-testid="update-meal">
        <AppBar position="static">
          <Toolbar>
            <CustomIconButton color="secondary" aria-label="Menu">
              <ArrowBack onClick={() => navigate('/')} />
            </CustomIconButton>
            <Logo src={logo} alt="logo" />
            <CustomTypography variant="h6" color="secondary">
              Update Meal
            </CustomTypography>
          </Toolbar>
        </AppBar>
        <Form>
          <FormControls>
            <InputWrapper>
              <CustomTextField
                placeholder="Meal name"
                value={name}
                name="name"
                onChange={e => this.onChangeHandler('name', e.target.value)}
              />
            </InputWrapper>
            <InputWrapper>
              <CustomTextField
                placeholder="Number of calories"
                value={calories}
                name="calories"
                onChange={e => this.onChangeHandler('calories', e.target.value)}
              />
            </InputWrapper>
            <InputWrapper>
              <CustomTextField
                placeholder="Date"
                value={`${date}T${time}`}
                type="datetime-local"
                InputLabelProps={{
                  shrink: true,
                }}
                onChange={e => this.onChangeHandler('datetime', e.target.value)}
              />
            </InputWrapper>
          </FormControls>
          {/* TODO: It should be disabled if no changes are made */}
          <Button
            variant="contained"
            color="primary"
            disabled={!change}
            name="submit-update-meal"
            onClick={() => {
              updateMeal(mealId, {
                name,
                calories,
                date,
                time,
              })
                .then(() => {
                  this.setState({
                    positiveSnackbarOpen: true,
                    positiveMessage: 'Meal updated successfully',
                  });
                  setTimeout(() => {
                    navigate('/');
                  }, 1000);
                }).catch(() => {
                  this.setState({
                    negativeSnackbarOpen: true,
                    negativeMessage: 'Couldn\'t update meals. Try again!',
                  });
                });
            }}
          >
            Update Meal
          </Button>
        </Form>
        <Snackbar
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
          open={positiveSnackbarOpen}
          autoHideDuration={1000}
          onClose={this.handlePositiveSnackbarClose}
        >
          <CustomSnackbar
            onClose={this.handlePositiveSnackbarClose}
            variant="success"
            message={positiveMessage}
          />
        </Snackbar>
        <Snackbar
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
          open={negativeSnackbarOpen}
          autoHideDuration={1000}
          onClose={this.handleNegativeSnackbarClose}
        >
          <CustomSnackbar
            onClose={this.handleNegativeSnackbarClose}
            variant="error"
            message={negativeMessage}
          />
        </Snackbar>
      </Wrapper>
    );
  }
}

UpdateMeal.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string.isRequired,
    role: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    dailyCaloriesLimit: PropTypes.number.isRequired,
  }),
  meals: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    time: PropTypes.string.isRequired,
    calories: PropTypes.number.isRequired,
  })),
  updateMeal: PropTypes.func.isRequired,
  getAllMeals: PropTypes.func.isRequired,
  navigate: PropTypes.func.isRequired,
  mealId: PropTypes.string.isRequired,
};

UpdateMeal.defaultProps = {
  user: null,
  meals: null,
};

const mapStateToProps = state => ({
  user: isEmpty(state.user.data) ? null : state.user.data,
  meals: state.meals.data,
});

const mapDispatchToProps = dispatch => ({
  getAllMeals: () => dispatch(fetchMealsActionCreator()),
  updateMeal: (mealId, updates) => dispatch(updateMealActionCreator(mealId, updates)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UpdateMeal);
