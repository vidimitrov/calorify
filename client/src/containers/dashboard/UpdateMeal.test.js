import React from 'react';
import renderer from 'react-test-renderer';
import { UpdateMeal } from './UpdateMeal';

it('UpdateMeal container renders correctly', () => {
  const props = {
    updateMeal: () => { },
    getAllMeals: () => { },
    navigate: () => { },
    account: {
      id: 'some-fake-id',
      name: 'John',
      dailyCaloriesLimit: 2000,
      role: 'user',
      email: 'john@email.com',
    },
    mealId: 'some-fake-meal-id',
    meals: [{
      id: 'some-fake-meal-id',
      name: 'some-name',
      date: '2018-11-10',
      time: '13:35',
      calories: 320,
    }],
  };
  const tree = renderer.create(
    <UpdateMeal {...props} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
