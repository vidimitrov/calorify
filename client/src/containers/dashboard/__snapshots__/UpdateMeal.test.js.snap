// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateMeal container renders correctly 1`] = `
<div
  className="Wrapper-m60uq5-0 hkWJUG"
>
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38"
    >
      <button
        aria-label="Menu"
        className="MuiButtonBase-root-47 MuiIconButton-root-41 MuiIconButton-colorSecondary-44 CustomIconButton-sc-1edavx9-0 eiNqgO"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-46"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-50"
            focusable="false"
            onClick={[Function]}
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-154"
        />
      </button>
      <img
        alt="logo"
        className="Logo-sc-1vcm3wd-0 lmrySJ"
        src="logo.png"
      />
      <h6
        className="MuiTypography-root-59 MuiTypography-h6-76 MuiTypography-colorSecondary-90 CustomTypography-e31nmz-0 kgGAHm"
      >
        Update Meal
      </h6>
    </div>
  </header>
  <div
    className="Form-ycfidt-0 jwaFmv"
  >
    <div
      className="FormControls-sc-7fqkjw-0 dcNCfX"
    >
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <div
            className="MuiInputBase-root-111 MuiInput-root-98 MuiInput-underline-102 MuiInputBase-formControl-112 MuiInput-formControl-99"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-121 MuiInput-input-106"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Meal name"
              required={false}
              type="text"
              value="some-name"
            />
          </div>
        </div>
      </div>
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <div
            className="MuiInputBase-root-111 MuiInput-root-98 MuiInput-underline-102 MuiInputBase-formControl-112 MuiInput-formControl-99"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-121 MuiInput-input-106"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Number of calories"
              required={false}
              type="text"
              value={320}
            />
          </div>
        </div>
      </div>
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <div
            className="MuiInputBase-root-111 MuiInput-root-98 MuiInput-underline-102 MuiInputBase-formControl-112 MuiInput-formControl-99"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-121 MuiInput-input-106 MuiInputBase-inputType-124 MuiInput-inputType-109"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Date"
              required={false}
              type="datetime-local"
              value="2018-11-10T13:35"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      className="MuiButtonBase-root-47 MuiButtonBase-disabled-48 MuiButton-root-128 MuiButton-contained-139 MuiButton-containedPrimary-140 MuiButton-raised-142 MuiButton-raisedPrimary-143 MuiButton-disabled-148"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-129"
      >
        Update Meal
      </span>
    </button>
  </div>
</div>
`;
